### PR TITLE
:sparkles: (youtube) insert view count into youtube

### PR DESCRIPTION
**Pull request by david30907d**
## Types of changes

- **New feature**

## Description

insert viewCount, likecount etc into BQ
## Checklist:

too lazy to add test case 😓 
- [ ] Add test cases to all the changes you introduce
- [x] Run `poetry run pytest` locally to ensure all linter checks pass
- [x] Update the documentation if necessary

## Steps to Test This Pull Request

1. follow readme to kick off a job on your local
2. go to bq and you should see incoming data
Screen Shot 2021-09-20 at 9 51 53 PM

## Expected behavior

ditto

## Additional context

<https://hackmd.io/54uxwADjSNaeDDMsdXOp3g>

### CODE REVIEW

File: dags/ods/youtube/dags/dag.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        python_callable=youtube_api.get_video_ids,
+        provide_context=True,
+    )
+
+    SAVE_STATISTICS_DATA_2_BQ = PythonOperator(
+        task_id="SAVE_STATISTICS_DATA_2_BQ",
+        python_callable=youtube_api.save_statistics_data_2_bq,
+        provide_context=True,
+    )
+    CREATE_TABLE_IF_NEEDED >> GET_VIDEO_IDS >> SAVE_STATISTICS_DATA_2_BQ
```
Comment (david30907d) [inline dags/ods/youtube/dags/dag.py:42]: main logic as follow:
1. create a bigquery table
2. get video ids from youtube
3. save statistics data of your video into bq

File: dags/ods/youtube/sqls/create_table.sql [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+(
+    created_at TIMESTAMP NOT NULL,
+    videoId STRING NOT NULL,
+    title STRING NOT NULL,
+    viewCount INT64 NOT NULL,
+    likeCount INT64 NOT NULL,
+    dislikeCount INT64 NOT NULL,
+    favoriteCount INT64 NOT NULL,
+    commentCount INT64 NOT NULL
+);
```
Comment (david30907d) [inline dags/ods/youtube/sqls/create_table.sql:11]: schema of this youtube table

File: dags/ods/youtube/udfs/youtube_api.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+import pandas as pd
+from airflow import macros
+from airflow.hooks.http_hook import HttpHook
+from airflow.models import Variable
+from google.cloud import bigquery
+from utils.hook_related import RETRY_ARGS
+
+# channel id of YouTube is public to everyone, so it's okay to commit this ID into git
+CHANNEL_ID = "UCHLnNgRnfGYDzPCCH8qGbQw"
```
Comment (david30907d) [inline dags/ods/youtube/udfs/youtube_api.py:12]: FYI

File: dags/ods/youtube/udfs/youtube_api.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+MAX_RESULTS = 50
+TABLE = f"{os.getenv('BIGQUERY_PROJECT')}.ods.ods_youtubeStatistics_videoId_datetime"
+
+
+def create_table_if_needed():
+    client = bigquery.Client(project=os.getenv("BIGQUERY_PROJECT"))
+    sql = Path("dags/ods/youtube/sqls/create_table.sql").read_text().format(TABLE)
+    client.query(sql)
```
Comment (david30907d) [inline dags/ods/youtube/udfs/youtube_api.py:20]: create table

File: dags/ods/youtube/udfs/youtube_api.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    response_json = http_conn.run_with_advanced_retry(
+        endpoint=base_url,
+        _retry_args=RETRY_ARGS,
+        headers={"Content-Type": "application/json", "Cache-Control": "no-cache"},
+    ).json()
+    video_metadatas += [
+        {"videoId": item["id"]["videoId"], "title": item["snippet"]["title"]}
+        for item in response_json["items"]
+        if "videoId" in item["id"]
+    ]
```
Comment (david30907d) [inline dags/ods/youtube/udfs/youtube_api.py:40]: this is how I get the metadata from Youtube

File: dags/ods/youtube/udfs/youtube_api.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    while response_json.get("nextPageToken"):
```
Comment (david30907d) [inline dags/ods/youtube/udfs/youtube_api.py:41]: iterate the pagination~

File: dags/ods/youtube/udfs/youtube_api.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+            endpoint=f"{base_url}&pageToken={next_page_token}",
+            _retry_args=RETRY_ARGS,
+            headers={"Content-Type": "application/json", "Cache-Control": "no-cache"},
+        ).json()
+        video_metadatas += [
+            {"videoId": item["id"]["videoId"], "title": item["snippet"]["title"]}
+            for item in response_json["items"]
+        ]
+    task_instance = context["task_instance"]
+    task_instance.xcom_push("GET_VIDEO_IDS", video_metadatas)
```
Comment (david30907d) [inline dags/ods/youtube/udfs/youtube_api.py:53]: pass `video_metadatas` to the next task, which is `save_statistics_data_2_bq`

File: dags/ods/youtube/udfs/youtube_api.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+
+
+def save_statistics_data_2_bq(**context):
+    def _init():
+        client = bigquery.Client(project=os.getenv("BIGQUERY_PROJECT"))
+        http_conn = HttpHook(method="GET", http_conn_id="youtube")
+        execution_date = context["execution_date"].replace(tzinfo=None)
+        task_instance = context["task_instance"]
+        video_metadatas = task_instance.xcom_pull("GET_VIDEO_IDS", key="GET_VIDEO_IDS")
```
Comment (david30907d) [inline dags/ods/youtube/udfs/youtube_api.py:62]: get `save_statistics_data_2_bq` from previous step

File: dags/ods/youtube/udfs/youtube_api.py [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+            video_id = video_metadata["videoId"]
+            title = video_metadata["title"]
+            response_json = http_conn.run_with_advanced_retry(
+                endpoint=f"/youtube/v3/videos?id={video_id}&key={Variable.get('YOUTUBE_KEY')}&part=statistics",
+                _retry_args=RETRY_ARGS,
+                headers={
+                    "Content-Type": "application/json",
+                    "Cache-Control": "no-cache",
+                },
+            ).json()
```
Comment (david30907d) [inline dags/ods/youtube/udfs/youtube_api.py:77]: this is how I get the statistics data

File: dags/ods/youtube/dags/dag.py [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+from airflow.operators.python_operator import PythonOperator
+from ods.youtube.udfs import youtube_api
+
+DEFAULT_ARGS = {
+    "owner": "davidtnfsh",
+    "depends_on_past": False,
+    "start_date": datetime(2021, 9, 19),
+    "retries": 2,
+    "retry_delay": timedelta(minutes=5),
+    "on_failure_callback": lambda x: "Need to send notification to Telegrame",
```
Comment (uranusjr) [inline dags/ods/youtube/dags/dag.py:16]:
```suggestion
    "on_failure_callback": lambda x: "Need to send notification to Telegram",
```

**Review by tai271828 — APPROVED** [review]
Some get method may be suggested to be wrapped in a try-except statement in case there is no data available for any reason. This kind of data sanity check could be implemented with another pull request.

File: dags/ods/youtube/dags/dag.py (hunk 10 above)
Comment (david30907d) [reply dags/ods/youtube/dags/dag.py:16]: good catch haha

**Comment by david30907d** [comment]
> Some get method may be suggested to be wrapped in a try-except statement in case there is no data available for any reason. This kind of data sanity check could be implemented with another pull request.

good to know, thx~

**Pull request merged**